### PR TITLE
Encode URL correctly for file upload

### PR DESCRIPTION
**Pull request by cyqsimon**
I noticed that in the current version, if you try to upload any file with a [reserved character](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURI#encodeuri_vs_encodeuricomponent) in its name, well, things break, badly.

As of "how badly?", in the worst case it can cause some weird-ass behaviour in the file system:

<hr>

### WARNING: Read procedure in full before trying

1. On Windows 10, run `httplz -p 80 -w ./testdir`
2. Prepare a file, with `_#` in its name (`_` here means <kbd>space</kbd>). E.g. `trouble #maker.test`
3. Upload using Web UI
4. Observe that the file name seems truncated at `_#`
5. Try deleting using Web UI. Unable to delete with response code `404`.
6. Try deleting on host using Windows' native file system. Unable to delete with error "Could not find this item".
7. Try deleting using cmd. `del <filename>` errors too.

To actually delete the file, I had to use [this method](https://www.winhelponline.com/blog/could-not-find-this-item-when-deleting-file-folder-windows/#cmd_sfn).

Things are not **as bad** on Linux. `rm` gets the job done, but still.

<hr>

So basically I changed some Javascript to make sure things are encoded correctly, and everything seems to be working. However, this also exposes a security/reliability issue of the serverside code, which is that it doesn't sanity-check the URL for proper encoding (alongside another few issues regarding path). I can open another PR to try to fix these serverside issues, but I might require assistance on that.

### CODE REVIEW

File: assets/upload.js [hunk 1]
```diff
@@ -59,18 +57,20 @@ window.addEventListener("load", function() {
...
+      });
     } else
-      entry.createReader().readEntries(function(e) {
-        e.forEach(function(f) {
-          recurse_upload(f, base_url)
+      entry.createReader().readEntries(function(e_arr) {
+        e_arr.forEach(function(f) {
+          recurse_upload(f, base_path)
         });
       });
```
Comment (cyqsimon) [inline assets/upload.js:75]: I'm pretty sure this original conditional logic is a screw-up.

Comment (nabijaczleweli) [reply assets/upload.js:75]: And I think it's correct because I wrote it that way. Some secrets cannot be known.

**Comment by nabijaczleweli** [comment]
I've worked around Firefox being overeager and trimming the paths by encoding the final character when needed (this fixes both opening and deletion), and fixed path chunks not being URL-encoded. Can you try `develop` (at least 3c781ab04c2cef47edb3f3dd0e2709ff92e8f978) on your end?

As for the files not being deletable with the default Windows userland, I couldn't care less. Works for me.
